### PR TITLE
Introduce nixpkgs-fmt to format nix sources

### PR DESCRIPTION
**Pull request by gilligan**
This PR introduces [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt) to format all nix sources for consistent indentation style that also optimizes for minimal diffs.

- Enables nixpkgs-fmt in the git pre-commit hook (introduced via #2398)
- Adds a CI check [nix/tests/nixpkgs-fmt.nix](./nix/tests/nixpkgs-fmt.nix)
- Adds an `.ignore` file for excluding files/patterns from re-formatting
- Re-formats all nix sources

### CODE REVIEW

**Comment by michaelpj** [comment]
I'm in favour. I looked at this before, and I think I wasn't sure which Nix formatter was winning, but `nixpkgs-fmt` seems to be pretty popular. And I don't see anything objectionable in the output.

Couple of things:
- Yes, we should have a CI check. You can have a look at `nix/tests/stylish-haskell.nix` as an example.
- It would be cool to have a pre-commit hook too, why not!
- The Nix code in `nix/stack.materialized` is *generated*, and the CI will check that the generated output matches, so we should *not* format it.

**Comment by michaelpj** [comment]
Oh, `nix/sources.nix` is also generated by `niv`, so that shouldn't be touched either. I wonder if we can use the stuff in `.gitattributes` :thinking:

**Comment by gilligan** [comment]
https://github.com/nix-community/nixpkgs-fmt/issues/191

I knew there was something but had to look through the repository first (Lacking some documentation for that!):

 Files/Directories can be excluded using `.ignore` files on the repository root with the same syntax as `.gitignore` files. I'll try that.

File: nix/sources.json [hunk 1]
```diff
@@ -48,6 +48,19 @@
     "url": "https://github.com/input-output-hk/iohk-nix/archive/24ddc8f531f320c9e2394d205d0df39857222901.tar.gz",
     "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
   },
+  "naersk": {
```
Comment (michaelpj) [inline nix/sources.json:51]: Why do we need this?

File: .ignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+nix/stack.materialized
+nix/sources.nix
```
Comment (michaelpj) [inline .ignore:2]: There still seem to be changes in `sources.nix`?

File: nix/overlays/nixpkgs-overrides.nix [hunk 3]
```diff
@@ -1,7 +1,9 @@
 self: super: {
+  naersk = super.callPackage ((import ../sources.nix).naersk) { };
```
Comment (michaelpj) [inline nix/overlays/nixpkgs-overrides.nix:2]: Ah I worked it out, it's needed for the build of `nixpkgs-fmt`, and it's all overlay magic!

File: nix/overlays/nixpkgs-overrides.nix [hunk 4]
```diff
@@ -1,7 +1,9 @@
...
+  nixpkgs-fmt = super.callPackage ((import ../sources.nix).nixpkgs-fmt) { };
```
Comment (michaelpj) [inline nix/overlays/nixpkgs-overrides.nix:3]: Ideally we should put comments in here saying when we can go back to the nixpkgs version - what's wrong with the nixpkgs version of nixpkgs-fmt? Just that it's too old? Should we go back when we update to 20.09?

**Review by michaelpj — APPROVED** [review]
Generally LGTM

File: nix/overlays/nixpkgs-overrides.nix (hunk 4 above)
Comment (gilligan) [reply nix/overlays/nixpkgs-overrides.nix:3]: yes it is unfortunately a bit old - upstream has already seen a 1.0 release that features some considerable changes in how the code is formatted.

I can (and probably should) add a comment though that we can switch to using nixpkgs-fmt from nixpkgs when switching to 20.09

File: .ignore (hunk 2 above)
Comment (gilligan) [reply .ignore:2]: uhm... yes, weird. Let me check that.

Comment (gilligan) [reply .ignore:2]: Fixed it.

File: nix/overlays/nixpkgs-overrides.nix (hunk 4 above)
Comment (gilligan) [reply nix/overlays/nixpkgs-overrides.nix:5]: Comment added, thanks for pointing that out.

**Pull request merged**